src/main/java/com/wultra/security/userdatastore/userclaims/EncryptionService.java

```diff
@@ -0,0 +1,160 @@
+        final byte[] claimsBytes = claims.getBytes(StandardCharsets.UTF_8);
+
+        try {
+            final byte[] iv = keyGenerator.generateRandomBytes(16);
+            final byte[] encrypted = aesEncryptionUtils.encrypt(claimsBytes, iv, secretKey);
+            final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            baos.write(iv);
+            baos.write(encrypted);
+            return Base64.getEncoder().encodeToString(baos.toByteArray());
+        } catch (GenericCryptoException | CryptoProviderException | InvalidKeyException | IOException e) {
+            throw new EncryptionException("Unable to decrypt claims for user ID: " + userId, e);
+        }
+    }
+
+    private String fromDBValue(final UserClaimsEntity entity) {
+        final String userId = entity.getUserId();
+        final SecretKey secretKey = fetchDerivedKey(userId);
+        final byte[] claimsBytes = Base64.getDecoder().decode(entity.getClaims());
+
+        if (claimsBytes.length < 16) {
+            throw new EncryptionException("Invalid encrypted private key format - the byte array is too short");
+        }
+
+        // IV is present in first 16 bytes
+        final byte[] iv = Arrays.copyOfRange(claimsBytes, 0, 16);
+
+        // Encrypted claims is present after IV
+        final byte[] encryptedClaims = Arrays.copyOfRange(claimsBytes, 16, claimsBytes.length);
+
+        try {
+            final byte[] decryptedClaims = aesEncryptionUtils.decrypt(encryptedClaims, iv, secretKey);
+            return new String(decryptedClaims);
+        } catch (InvalidKeyException | GenericCryptoException | CryptoProviderException e) {
+            throw new EncryptionException("Unable to decrypt claims for user ID: " + userId, e);
+        }
+    }
+
+    private SecretKey fetchDerivedKey(final String userId) {
+        if (!StringUtils.hasText(masterDbEncryptionKeyBase64)) {
+            throw new EncryptionException("masterDbEncryptionKey is not configured");
+        }
+
+        final SecretKey masterDbEncryptionKey = convertBytesToSharedSecretKey(Base64.getDecoder().decode(masterDbEncryptionKeyBase64));
+        return deriveSecretKey(masterDbEncryptionKey, userId);
+    }
+
+    private static SecretKey convertBytesToSharedSecretKey(final byte[] bytesSecretKey) {
+        return new SecretKeySpec(bytesSecretKey, "AES");
+    }
+
+    /**
+     * Derive secret key from master DB encryption key and user ID.
+     *
+     * @param masterDbEncryptionKey Master DB encryption key.
+     * @param userId User ID, used as index for KDF_INTERNAL
+     * @return Derived secret key.
+     * @see <a href="https://github.com/wultra/powerauth-server/blob/develop/docs/Encrypting-Records-in-Database.md">Encrypting Records in Database</a>
+     */
+    private SecretKey deriveSecretKey(final SecretKey masterDbEncryptionKey, final String userId) {
+        final byte[] index = userId.getBytes(StandardCharsets.UTF_8);
```

Comment:
Do we want to check the `userId` for nullability? In case of a bug, all keys could be derived from a `null` value, making the encryption quite weak.

Reply:
`userId` is a primary key and we use bean validation, but you are right, for encryption, we should check again.